gen_funcs.sh: check_disk_space_requirements(): Use stat(1) to get free disk space

vrza opened the pull request:
By using `stat(1)` to get free disk space we avoid tricky parsing of `df(1)` output.

Original code is brittle:
- it uses `df -BM` to get free space in megabytes
- parses out `df(1)` output to get a substring like `"1200M"`
- attempts to parse this substring in order to remove the suffix `"M"` and get the numeric value

The problem arises because the suffix (`"M"`) is subject to localization, but parsing looks for a latin letter `"M"`. 

Forcing `LC_ALL=C` to get more predictable output from `df(1)` was added as a workaround.

By using `stat(1)` we avoid this complication, as we don't need to parse its output (and hence we don't need to deal with localization either).

Karlson2k commented on gen_funcs.sh:
```diff
@@ -1853,20 +1853,21 @@ check_disk_space_requirements() {
 		fi
 
-		available_free_disk_space=$(unset POSIXLY_CORRECT && LC_ALL="C" df -BM "${BOOTDIR}" | awk '$3 ~ /[0-9]+/ { print $4 }')
+		available_free_disk_space=$(expr "$(stat -fc%f "${BOOTDIR}")" \* "$(stat -fc%S "${BOOTDIR}")")
```
``` C
available_free_disk_space=$(($(stat -fc '%f*%S' "${BOOTDIR}")))
````

vrza replied:
Yeah, since we are using GNU bash as the interpreter, that works and helps make a single call to `stat(1)` insead of two.

My programming style is geared towards code portability, hence I use the POSIX shell syntax.

Since both variants are correct and the difference in script execution time is negligible, I'll defer to the maintainer on the choice of syntax.

Karlson2k replied:
I usually prefer to write scripts in a portable way, but since genkernel is tied to bash, why not to use bash advantages?

BTW, `$(())` mathematics is not a bashism, it is standard POSIX:
https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_04

vrza replied:
Thanks for pointing out that `$(())` arithmetic expansion is in fact POSIX. I will update the PR.

> since genkernel is tied to bash, why not to use bash advantages?

In general, while genkernel as a whole is tied to bash, we might want to reuse parts of it in other scripts. Checking for free disk space seems like something that can be useful even outside of genkernel context.

vrza committed:
gen_funcs.sh: use stat(1) to check free disk space

By using stat(1) to get free disk space we avoid tricky parsing
of df(1) output.

Signed-off-by: Vladimir Vrzić <vvrzic@gmail.com>

vrza commented:
Resolved conflicts with latest upstream `master` branch.

thesamesam commented:
First, I'm sorry this has taken ages to get to.

Would you mind commenting your `Signed-off-by: name <email>` (and possibly amending your commit with it), then I'll get this in?

I definitely agree `stat` is the right tool for the job here and is far more robust.

vrza commented:
Sure, changed commit message.

thesamesam commented:
Thank you!
